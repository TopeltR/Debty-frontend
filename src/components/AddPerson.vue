<template>
    <b-row>
        <b-col cols='8' col-md="9">
            <autocomplete id='user' v-model='user' :placeholder='"Name"' :field='field'
                          :items='state.allPeople'
                          :key-extractor='getFullName'></autocomplete>
        </b-col>
        <b-col cols='4' col-md="3">
            <button type='button' v-on:click='addPerson'
                    class='btn btn-primary wide'>
                Add
            </button>
        </b-col>
    </b-row>
</template>

<script>
    import Autocomplete from '@/components/Autocomplete.vue';
    import userStore from '../stores/UserStore.ts';

    export default {
        name: "AddPerson",
        components: {Autocomplete},
        props: {
            state: {
                type: Object,
                allPeople: {
                    type: Array,
                },
                people: {
                    type: Array,
                },
            },
        },
        data: () => ({
            user: {},
            field: {value: ''},
        }),
        mounted() {

        },
        methods: {
            getFullName(user) {
                return user.firstName + ' ' + user.lastName;
            },
            addPerson() {
                userStore.getUser().then((user) => {
                    if (this.user && this.user.firstName && !this.state.people.map(user => user.email).includes(this.user.email)
                        && this.user.email !== user.email) {
                        this.state.people.push(this.user);
                        this.state.allPeople = this.state.allPeople.filter((person) => !this.state.people.includes(person));
                        this.field.value = '';
                    }
                });
            },
        }
    }
</script>

<style scoped>
    .wide {
        width: 100%;
    }
</style>
