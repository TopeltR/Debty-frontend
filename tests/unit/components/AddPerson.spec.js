import AddPerson from '@/components/AddPerson';
import userStore from '@/stores/UserStore'
import {mount} from "@vue/test-utils";
import * as Vue from "vue";
import BootstrapVue from 'bootstrap-vue';

describe('AddPerson.vue', () => {
    Vue.use(BootstrapVue);

    const user = {email: 'test@test.ee'};
    userStore.getUser = jest.fn(() => Promise.resolve(user));

    describe('addPerson', () => {
        it('should add user to people and remove user from allPeople and empty field.value', async () => {
            const addPerson = mount(AddPerson, {
                propsData: {
                    state: {
                        allPeople: [],
                        people: []
                    }
                }
            });
            addPerson.user = {firstName: 'name', email: 'nottest@test.ee'};

            await addPerson.vm.addPerson();

            expect(addPerson.vm.state.people.includes(user))
        })
    });

    it('should render', () => {
        const addPerson = mount(AddPerson, {
            propsData: {
                state: {
                    allPeople: [],
                    people: []
                }
            }
        });
        expect(addPerson.html()).toMatchSnapshot();
    })
});
