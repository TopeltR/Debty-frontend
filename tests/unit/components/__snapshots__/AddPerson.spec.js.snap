// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPerson.vue should render 1`] = `
<div class="row">
  <div col-md="9" class="col-8">
    <div class="autocomplete" id="user" value="[object Object]"><input type="text" placeholder="Name" class="form-control">
      <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
    </div>
  </div>
  <div col-md="3" class="col-4"><button type="button" class="btn btn-primary wide">
      Add
    </button></div>
</div>
`;
